jlm/llvm/ir/RvsdgModule.hpp

```diff
@@ -14,70 +14,57 @@
 namespace jlm::llvm
 {
 
-/* impport class */
-
-class impport final : public jlm::rvsdg::impport
+/**
+ * Represents an import into the RVSDG of an external entity.
+ * It is used to model LLVM module declarations.
+ */
+class GraphImport final : public rvsdg::GraphImport
 {
-public:
-  virtual ~impport();
-
-  impport(
-      std::shared_ptr<const jlm::rvsdg::valuetype> valueType,
-      const std::string & name,
-      const linkage & lnk)
-      : jlm::rvsdg::impport(PointerType::Create(), name),
-        linkage_(lnk),
+private:
+  GraphImport(
+      rvsdg::graph & graph,
+      std::shared_ptr<const rvsdg::valuetype> valueType,
+      std::string name,
+      llvm::linkage linkage)
+      : rvsdg::GraphImport(graph, PointerType::Create(), std::move(name)),
+        Linkage_(std::move(linkage)),
         ValueType_(std::move(valueType))
   {}
 
-  impport(const impport & other) = default;
-
-  impport(impport && other) = default;
-
-  impport &
-  operator=(const impport &) = delete;
-
-  impport &
-  operator=(impport &&) = delete;
-
-  const jlm::llvm::linkage &
-  linkage() const noexcept
+public:
+  [[nodiscard]] const linkage &
+  Linkage() const noexcept
   {
-    return linkage_;
+    return Linkage_;
   }
 
   [[nodiscard]] const std::shared_ptr<const jlm::rvsdg::valuetype> &
-  Type() const noexcept
+  ValueType() const noexcept
   {
     return ValueType_;
   }
 
-  [[nodiscard]] const jlm::rvsdg::valuetype &
-  GetValueType() const noexcept
+  GraphImport &
+  Copy(rvsdg::region & region, rvsdg::structural_input * input) override;
+
+  static GraphImport &
```

Comment:
This is the only part that I find a bit odd, returning a reference -- I would have found it more natural to return a pointer: catching a returned pointer with "auto x = GraphImport::Create(...)" does the right thing, while it will do totally the wrong thing when returning a reference.

Reply:
> catching a returned pointer with "auto x = GraphImport::Create(...)" does the right thing, while it will do totally the wrong thing when returning a reference.

What do you mean by "will do totally the wrong thing"?

Instead of `auto x = ...`, you do now `auto & x = ...`. I do not see the problem? The copy constructor and copy assignment operator are deleted from output (and therefore argument). Thus, it is also not possible to accidentally do `auto x = ...` and create a unwanted copy. The benefit of a reference is that the user knows a initialized argument is returned (instead of a potential nullptr).

Reply:
okay relying on deleted copy constructor is safe indeed